pull/github.go

```diff
@@ -260,10 +267,85 @@ func (ghc *GitHubContext) Reviews() ([]*Review, error) {
 	return ghc.reviews, nil
 }
 
+func (ghc *GitHubContext) ListRepositoryCollaborators() (map[string]string, error) {
+	if ghc.collaborators == nil {
+		if err := ghc.loadPagedData(); err != nil {
+			return nil, err
+		}
+	}
+	return ghc.collaborators, nil
+}
+
+func (ghc *GitHubContext) ListOrganizationMembers(org string) ([]string, error) {
+	opt := &github.ListMembersOptions{
+		ListOptions: github.ListOptions{
+			PerPage: 100,
+		},
+	}
+
+	// get all pages of results
+	var allUsers []string
+
+	for {
+		users, resp, err := ghc.client.Organizations.ListMembers(ghc.ctx, org, opt)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list members of org %s page %d", org, opt.Page)
+		}
+		for _, u := range users {
+			allUsers = append(allUsers, u.GetLogin())
+		}
+		if resp.NextPage == 0 {
+			break
+		}
+		opt.Page = resp.NextPage
+	}
+
+	return allUsers, nil
+}
+
+func (ghc *GitHubContext) ListTeamMembers(org, teamName string) ([]string, error) {
+	opt := &github.TeamListTeamMembersOptions{
+		ListOptions: github.ListOptions{
+			PerPage: 100,
+		},
+	}
+
+	team, _, err := ghc.client.Teams.GetTeamBySlug(ghc.ctx, org, teamName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Unable to get information for team %s/%s", org, team)
+	}
+
+	// get all pages of results
+	var allUsers []string
+
+	for {
+		users, resp, err := ghc.client.Teams.ListTeamMembers(ghc.ctx, team.GetID(), opt)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list team %s/%s members page %d", org, teamName, opt.Page)
+		}
+		for _, u := range users {
+			allUsers = append(allUsers, u.GetLogin())
+		}
+		if resp.NextPage == 0 {
+			break
+		}
+		opt.Page = resp.NextPage
+	}
+
+	return allUsers, nil
+}
+
 func (ghc *GitHubContext) loadPagedData() error {
 	// this is a minor optimization: make max(c,r) requests instead of c+r
 	var q struct {
 		Repository struct {
+			Collaborators struct {
+				PageInfo v4PageInfo
+				Edges    []struct {
+					Permission githubv4.RepositoryPermission `graphql:"permission"`
+					Node       v4Collaborator                `graphql:"node"`
+				}
+			} `graphql:"collaborators(first: 100, after: $collaboratorCursor)"`
```

Comment:
While including this as part of the "paged data" may reduce individual network calls, it increases the complexity of a GraphQL request that is called for _every_ PR evaluation. My instinct is that collaborators should be loaded on-demand, since the review request feature is only enabled for specific events and the review request feature is opt-in.

For the second point, I can be convinced that usage of review requests will be more widespread than I expect right now, but it still seems wasteful to load these for every comment or commit evaluation.

Reply:
I think you're right about the increasingly query cost, and I'll move it.